scripts/core/exp-command-buffer.yml

```diff
@@ -199,7 +226,7 @@ members:
       desc: "[in][optional][range(0, newWorkDim)] Array of newWorkDim unsigned values that describe the number of global work-items."
     - type: "size_t*"
       name: pNewLocalWorkSize
-      desc: "[in][optional][range(0, newWorkDim)] Array of newWorkDim unsigned values that describe the number of work-items that make up a work-group. If newWorkDim is non-zero and pNewLocalWorkSize is nullptr, then runtime implementation will choose the work-group size. If newWorkDim is zero and pNewLocalWorkSize is nullptr, then the local work size is unchanged."
+      desc: "[in][optional][range(0, newWorkDim)] Array of newWorkDim unsigned values that describe the number of work-items that make up a work-group. If pNewLocalWorkSize is nullptr, then the local work size is unchanged."
```

Comment:
As discussed on the standup, I'm unsure about what is the most intuitive behaviour for local work-size. It used to be that:

- `Work-dim == 0 && pNewLocalWorkSize == nullptr` - No update
- `Work-dim > 0 && pNewLocalWorkSize == nullptr` - Implementation defined local work-dim
- `Work-dim > 0 && pNewLocalWorkSize != nullptr` - User defined local work-dim

To make the api less confusing after the kernl binary update changes, I made it an error to have work-dim == 0. 
So the new behaviour is:

- `Work-dim == 0 && pNewLocalWorkSize == nullptr` - Error
- `Work-dim > 0 && pNewGlobalWorkSize != nullptr && pNewLocalWorkSize == nullptr` - Implementation defined local work-dim
- `Work-dim > 0 &&  && pNewGlobalWorkSize == nullptr  && pNewLocalWorkSize == nullptr` - No update
- `Work-dim > 0 && pNewLocalWorkSize != nullptr` - User defined local work-dim

We are losing the possibility of not updating the global work size and allowing the implementation to generate a local worksize. This seems like a very niche usecase. So maybe that's okay?

To me, this function has too many optional behaviours which makes it confusing. Anyway, would welcome feedback about which behaviour is more intuitive. Can go back to allowing dim == 0 if we think that's for the best. However that will add a few more verbose errors in the spec.

Reply:
Another option that was discussed is to allow all parameters (including newWorkDim) to be optional.

Essentially the implementation wouldn't do as many sanity checks on what the user provides. For example, if the user wants to change the work dimensions of a kernel node, that would be allowed. 

This approach could have the benefit of allowing the specification to be more consistent and easier to interpret due to less edge cases that need to return error codes.

- A point that still needs to be discussed is whether it can make sense to update the global work size without updating the local work-size (and vice-versa).
- Also unsure if allowing implementation defined local-work sizes on updates has a benefit.

Reply:
Thinking in terms of the sycl usage, there is `update_nd_range` where there is both local & global sizes, and `update_range` which is only global size. There is no way to set the local size without global size too.

If a user does `update_range()` I think it would be good to get a new implementation defined local work size. The user doesn't have the means to change this themselves. i.e. prefer option

```quote 
    newWorkDim == 0 && pNewLocalWorkSize == nullptr - Error
    newWorkDim > 0 && pNewGlobalWorkSize != nullptr && pNewLocalWorkSize == nullptr - Implementation defined local work-dim
    newWorkDim > 0 &&  pNewGlobalWorkSize == nullptr  && pNewLocalWorkSize == nullptr - No update
    newWorkDim > 0 && pNewLocalWorkSize != nullptr - User defined local work-dim
```


> Another option that was discussed is to allow all parameters (including newWorkDim) to be optional.

I'm not sure there is less edge cases, just edge cases in different places. e.g. what would the new value of nd-range be if you increased of decreased the work dim without updating that. How many elements would there be in the nd-range arrays if you didn't have to pass a work dim. I think these are solvable, but I think the least confusing API is one where everything is mandatory (apart from local work size, which can be nullptr for sycl::range kernels and stays null for all updates).

```quote 
    newWorkDim == 0 - Error
   newWorkDim > 0 &&  pNewGlobalWorkSize == nullptr - Error (global size not explicitly set)
   newWorkDim > 0 && pNewGlobalWorkSize != nullptr && pNewLocalWorkSize == nullptr - Implementation defined local work-dim
   newWorkDim > 0 && pNewGlobalWorkSize != nullptr  && pNewLocalWorkSize != nullptr - User defined local work-dim
```

Reply:
As discussed offline, I updated the spec to go mostly for the option that Ewan suggested with the exception that updating dimensions is also allowed as long as the global work size and the global work offset are updated at the same time